torch/nn/parallel/distributed.py

```diff
@@ -749,6 +749,17 @@ def __init__(
         else:
             self.process_group = process_group
 
+        if (
+            dist.get_backend(self.process_group) == "nccl"
+            and torch.cuda.nccl.version() >= (2, 17)
+        ):
+            # Note: NVIDIA recommends using CGA Cluster Size of 2 when using DDP.
+            default_cga = dist.ProcessGroupNCCL.Options().config.cga_cluster_size # type: ignore[attr-defined]
+            default_pg_nccl = self.process_group._get_backend(torch.device("cuda"))
+            current_cga = default_pg_nccl.options.config.cga_cluster_size
+            if current_cga == default_cga:
+                default_pg_nccl.options.config.cga_cluster_size = 2
+
```

Comment:
I would appreciate more information why 2 is recommended.
(The main reason being that it seems to be different from the default values picked by NCCL -- either 0 or 4 depending on SM arch:
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/types.html#c.cgaClusterSize)

Since this part touches DDP performance, I wonder if it would be easier for this PR to land if we move it into a separate PR?

Reply:
These numbers are coming from internal studies. Specifically, we found that using a NCCL CGA group of 2 gives the best performance when overlapped with GEMMs, whereas a NCCL CGA group of 4 is best in the absence of overlap. 

Moving this to a separate PR sounds good to me.

Reply:
I see, thanks for sharing the information and thanks for agreeing to move it into a separate PR!

I guess a follow-up question for that separate PR would be:
Since the process group is created outside, would there be any effect to modify the CGA setting here?
Are we assuming that the process group hasn't run any collective before being passed to DDP (hence NCCL communicator hasn't been created)?

Reply:
Removed the DDP related changes in this PR. Will introduce the default value of 2 for cga size and continue the discussion in a new PR.